### PR TITLE
extract generators, merge spans, language matching, docstring miscellany

### PR DESCRIPTION
**Pull request by bdewilde**
Grab-bag of changes and updates, most notably the following:

- Most of the `extract.py` functions are now generators (they yield one item at a time) rather than returning lists, with an eye towards performance on _very large_ documents. Tests were updated for the change. Because of caching, calls to these functions via the `TextDoc` class had to be wrapped in `list()`, which negates the aforementioned performance consideration. I may end up dropping the `@cachemethod` decorators...
- Both `TextDoc` and `TextCorpus` now check to make sure that the language of their spacy pipeline is the same as the language of the text(s) run through it. There is a way around the checks, but only through active user error, which isn't something I care to worry about.
- Spacy span merging now updated to match Spacy's latest. `TextDoc`s have a `.merge()` method; using it clears the cache, since the underlying data has changed.
- `extract.noun_phrases()` got renamed to `extract.noun_chunks()` to match Spacy's API. This is backwards incompatible, but this is Beta, so...

@danvalente, I choose you!

### CODE REVIEW

File: tests/test_extract.py [hunk 1]
```diff
@@ -13,220 +16,203 @@ def setUp(self):
...
-            "There's this notion that the wealthy use their money to buy politicians; more accurately, it's that they can buy policy, and specifically, tax policy," said Jared Bernstein, a senior fellow at the left-leaning Center on Budget and Policy Priorities who served as chief economic adviser to Vice President Joseph R. Biden Jr. "That's why these egregious loopholes exist, and why it's so hard to close them."
+            Two weeks ago, I was in Kuwait participating in an I.M.F. seminar for Arab educators. For 30 minutes, we discussed the impact of technology trends on education in the Middle East. And then an Egyptian education official raised his hand and asked if he could ask me a personal question: "I heard Donald Trump say we need to close mosques in the United States," he said with great sorrow. "Is that what we want our kids to learn?"
```
Comment (danvalente) [inline tests/test_extract.py]: Just curious...why the change of the text for this test?

File: textacy/extract.py [hunk 2]
```diff
@@ -168,35 +173,38 @@ def named_entities(doc,
...
-    return list(nes)
+    for ne in nes:
+        yield ne
 
 
-def noun_phrases(doc, drop_determiners=True, min_freq=1):
+def noun_chunks(doc, drop_determiners=True, min_freq=1):
     """
-    Extract an ordered list of noun phrases from a spacy-parsed doc, optionally
+    Extract an ordered sequence of noun chunks from a spacy-parsed doc, optionally
```
Comment (danvalente) [inline textacy/extract.py]: What's the difference between a noun phrase and a noun chunk?

File: tests/test_extract.py (hunk 1 above)
Comment (bdewilde) [inline tests/test_extract.py]: Too long, and I was only using the first part of it, anyway. Plus: Now we're testing on Thomas Friedman! :wink:

File: textacy/extract.py (hunk 2 above)
Comment (bdewilde) [inline textacy/extract.py]: No difference AFAIK, this was only to match up with Spacy's API.

File: textacy/spacy_utils.py [hunk 3]
```diff
@@ -90,28 +90,15 @@ def normalized_str(token):
...
 
 
-def merge_spans(spans, doc):
+def merge_spans(spans):
     """
-    Merge spans so that each only takes up a single token. Since each Span is
-    a view into the doc, merging one invalidates the others. (This will get fixed!)
-    The temporary solution is to gather the information first, before merging.
-    NB: This modifies `doc` in-place!
+    Merge spans *in-place* within parent doc so that each takes up a single token.
```
Comment (danvalente) [inline textacy/spacy_utils.py]: qq: So let's say I have a span `The quick brown fox ate a delicious white rabbit`, and I want to remove adjectives.  If I understand things, this function allows me to merge the resulting spans into `The fox ate a rabbit.'   But does merge keep the syntactic dependencies of the initial span, or does it reparse?

Comment (bdewilde) [inline textacy/spacy_utils.py]: No. A span is a consecutive sequence of tokens, and this function allows you to merge such sequences into a single token. So, Span("the quick brown fox") = [Token("the"), Token("quick"), Token("brown"), Token("fox")] becomes Token("the quick brown fox").

File: textacy/texts.py [hunk 4]
```diff
@@ -30,30 +33,31 @@ class TextDoc(object):
...
-            if spacy_pipeline is None else spacy_pipeline
+        self.lang = text_utils.detect_language(text) if lang == 'auto' else lang
+        if spacy_pipeline is None:
+            self.spacy_pipeline = data.load_spacy_pipeline(lang=self.lang)
+        else:
+            # check for match between text and supplied spacy pipeline language
+            if spacy_pipeline.lang != self.lang:
+                msg = 'TextDoc.lang {} != spacy_pipeline.lang {}'.format(
+                    self.lang, spacy_pipeline.lang)
+                raise ValueError(msg)
```
Comment (danvalente) [inline textacy/texts.py]: Nice add.

File: textacy/spacy_utils.py (hunk 3 above)
Comment (danvalente) [inline textacy/spacy_utils.py]: ah. got it.

**Comment by danvalente** [comment]
Pretty straight-forward change.  :shipit:
